File: go/036.go
func isValidSudoku(board [][]byte) bool {
    if len(board) == 0 || len(board[0]) == 0 {
        return false
    }
    m, n := len(board), len(board[0])
    row, col, cell := make([][]bool, m), make([][]bool, m), make([][]bool, m)
    for i := 0; i < m; i++ {
        row[i], col[i], cell[i] = make([]bool, n), make([]bool, n), make([]bool, n)

    }
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if board[i][j] == '.' {
                continue
            }
            c := int(board[i][j]-'0') - 1
            if row[i][c] || col[c][j] || cell[3*(i/3)+j/3][c] {
                return false
            }
            row[i][c], col[c][j], cell[3*(i/3)+j/3][c] = true, true, true
        }
    }
    return true
}

File: cpp/036.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<vector<bool>> row(9, vector<bool>(9));
        vector<vector<bool>> col(9, vector<bool>(9));
        vector<vector<bool>> cell(9, vector<bool>(9));
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.')
                    continue;
                int c = board[i][j] - '1';
                if (row[i][c] || col[c][j] || cell[3 * (i / 3) + j / 3][c])
                    return false;
                row[i][c] = col[c][j] = cell[3 * (i / 3) + j / 3][c] = true;
            }
        }
        return true;
    }
};
